### PR TITLE
[LF] clean value proto for LF 2.0

### PR DESCRIPTION
**Pull request by remyhaemmerle-da**
This is part of the effort #18457

### CODE REVIEW

File: daml-lf/transaction/src/main/protobuf/com/daml/lf/value.proto [hunk 1]
```diff
@@ -21,120 +21,114 @@ message VersionedValue {
...
 
+// Each component of the module name and the name have additional
+// lexical restrictions besides non-emptiness -- see the value.rst
+// spec for more info.
+message Identifier {
+  string package_id = 1; // non-empty
+  repeated string module_name = 2; // at least length 1
+  repeated string name = 3; // at least length 1
+}
+
```
Comment (remyhaemmerle-da) [inline daml-lf/transaction/src/main/protobuf/com/daml/lf/value.proto:32]: just move it up.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala [hunk 2]
```diff
@@ -214,79 +232,68 @@ object ValueCoder {
...
           case proto.Value.SumCase.DATE =>
             val d = Time.Date.fromDaysSinceEpoch(protoValue.getDate)
             d.fold(e => throw Err("error decoding date: " + e), ValueDate)
           case proto.Value.SumCase.TIMESTAMP =>
             val t = Time.Timestamp.fromLong(protoValue.getTimestamp)
             t.fold(e => throw Err("error decoding timestamp: " + e), ValueTimestamp)
+          case proto.Value.SumCase.NUMERIC =>
+            Numeric
+              .fromString(protoValue.getNumeric)
+              .fold(e => throw Err("error decoding decimal: " + e), ValueNumeric)
```
Comment (paulbrauner-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala:253]: Should it say "decoding numeric" instead?

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala [hunk 3]
```diff
@@ -206,6 +221,9 @@ object ValueCoder {
       if (version >= minVersion)
         throw Err(s"$description is not supported by transaction version $version")
 
+    def assertNoUnknownFields(msg: com.google.protobuf.Message) =
+      ensureNoUnknownFields(msg).left.foreach(e => throw Err("error decoding decimal: " + e))
```
Comment (paulbrauner-da) [inline daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala:225]:
```suggestion
      ensureNoUnknownFields(msg).left.foreach(e => throw Err("error decoding value: " + e))
```

**Review by paulbrauner-da — APPROVED** [review]
Thanks!

File: daml-lf/spec/value.rst [hunk 4]
```diff
@@ -258,287 +319,197 @@ otherwise try to compensate for "bad" input when reading decimal
...
+.. note: *this section is non-normative*
 
-* The first character must be ``$``, ``_``, or an ASCII letter;
-* Every other character must be ``$``, ``_``, an ASCII letter, or an
-  ASCII digit.
+    Every member of ``elements`` must conform to the same type.
 
-message Variant
-^^^^^^^^^^^^^^^
+message GenMap.Entry
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/spec/value.rst:373]: Good catch.

File: daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala (hunk 3 above)
Comment (remyhaemmerle-da) [reply daml-lf/transaction/src/main/scala/com/digitalasset/daml/lf/value/ValueCoder.scala:225]:
```suggestion
      ensureNoUnknownFields(msg).left.foreach(e => throw Err(e))
```

File: daml-lf/spec/value.rst [hunk 5]
```diff
@@ -146,24 +146,58 @@ Version history
...
+As of version 10, these fields are included, all required to be
+non-empty:
+
+* ``string`` package_id
+* repeated ``string`` module_name
+* repeated ``string`` name
+
+
+``package_id`` is a Daml-LF package ID, indicating the LF package in
+which the type is defined. package ID are restricted to be a
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/spec/value.rst:184]: :+1:

File: daml-lf/spec/value.rst [hunk 6]
```diff
@@ -146,24 +146,58 @@ Version history
...
+
+
+message Identifier
+^^^^^^^^^^^^^^^^^^
+
+(*since version 21*)
+
+A reference to a Daml-LF record or variant type.
+
+As of version 10, these fields are included, all required to be
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/spec/value.rst:175]: :+1:

File: daml-lf/spec/value.rst [hunk 7]
```diff
@@ -200,44 +234,71 @@ message Value
 
 An actual Daml-LF *serializable value*.
 
-(*since version 14*)
+(*since version 21*)
```
[the remark this answers is not in the document]
Comment (remyhaemmerle-da) [reply daml-lf/spec/value.rst:237]: Good idea.

**Pull request merged**